### ISSUE TITLE
Futures - Trade Contract Atomic Tx

### ISSUE THREAD

**Issue by patrickdugan** [OmniLayer/omnicore#378]
The trade contract tx type differs from the Dex transaction types in three key respects:

1) the transaction involves only changing the balance of a single smart property type (the contract in question) and is thus simpler so the 4 trade types aren't required.

2) instead of pushing reserved balances between addresses when orders match, the collateral property that backs the contract is pushed into reserve status or taken out of reserve status when the absolute value of the contract balance goes up or down.  

3) Reserve balances are only pushed at settlement time, not when trades clear. Instead reserve balances that are lost due to negative contract trade price differentials (i.e buying high and then selling low) are removed from the total balance calculation used to show to users their available totals, and only removed from the address when the contract expires and the differences are pushed out to the addresses of those with a net win. 

When a Trade Contract tx is formed for broadcast, in order be valid the number of contracts traded times the collateral requirement for the reserve smart property must be greater than the available balance of reserve asset (#contract \* collateral per contract >= balance of collateral) e.g the OMNI/USDT contract requires 5 OMNI per contract, the trade is for 10 contracts, but the OMNI balance is 40 contracts, the trade is invalid.  

If a transaction is accepted, all the units required to margin the order are held in reserve. 

If a transaction matches to another trade, then the trade executes, the balance of the contract is raised or reduced by 1 for a buy/sell, with the supply of contracts changing depending on the situation. There are 3 possible trades:

1) Both parties are opening contracts, new contracts come into existence and at settlement the reserve balances of the losing side will push to the address of the winning side. 

2) Both parties are closing contracts, the number of existing contracts reduces and as above the reserve balance of the losing side will push to the address of the winning side at settlement. 

3) One party is reducing the absolute value of their contract balance while another increases theirs. This would be effectively like someone trading off existing supply of contracts to someone else. However the atomic transaction is still considered unique so that at settlement the net transfers are calculated. 

Each trade encodes the trade price into its transaction meta-data for reference during settlement. Optionally, we can create an id number that has a unix-style timestamp included to index all trades in a chronology.

When someone reduces the absolute value of their contract balance, the trade price for the contract(s) used to do that is immediately taken and subtracted from the trade price of the contracts that opened those trades. The contracts with the prices that are closest to the market are taken into consideration first when calculating. The difference, is resulting in a loss, locks up collateral until expiration and removes it from both reserve and available balances for UI purposes.

**Comment by marvgmail** [comment]
@patrickdugan this looks pretty complex, with (implied) side effects. Can you dumb it down for those of us who aren't experts in the domain, but are able to follow - and code - logic flows? Simple, concise examples always help me. I think we also need a glossary so there's no misinterpretation of the terms you're using. Thanks.

**Comment by zathras-crypto** [comment]
I've suggested some flow charts showing what actions would be taken in what circumstances could help us to start translating this into something workable spec-wise.

Alternatively/additionally psuedo-code might work.

Once we can get that 'clear picture' I think we'll be able to start hammering out the technicalities.

**Comment by patrickdugan** [comment]
> 3. Reserve balances are only pushed at settlement time, not when trades clear. Instead reserve balances that are lost due to negative contract trade price differentials (i.e buying high and then selling low) are removed from the total balance calculation used to show to users their available totals, and only removed from the address when the contract expires and the differences are pushed out to the addresses of those with a net win

I want to update this earlier notion - a cache of funds that have been lost to closed-out negative trades should be kept in its own pocked dimension like the OmniDex fee cache. Funds would then be deducted from the cache to credit winning trades as they are closed out, on a first-come-first-serve queue basis.

**Comment by dexX7** [comment]
I think it's better to keep track of the specification or feature proposals in the spec repository:

https://github.com/OmniLayer/spec

Please feel free to open a new issue over there.
